pkg/tests/tasks/security/certificate/external_ca_test.go

```diff
@@ -16,101 +16,116 @@ package certificate
-			util.Inspect(err, "Failed to parse 'openssl s_client' output", "", t)
-			_, err = util.ShellMuteOutput(`awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "%s/proxy-cert-" counter ".pem"}' < %s/certs.pem`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to split certs into separate files", "", t)
-
-			// Compare them with the original certs
-			log.Log.Info("Verifying the root certificate")
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s -text -noout > %s/root-cert.crt.txt`, sampleCARoot, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s/proxy-cert-3.pem -text -noout > %s/pod-root-cert.crt.txt`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			if err := util.CompareFiles(fmt.Sprintf("%s/root-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-root-cert.crt.txt", tmpDir)); err != nil {
-				t.Errorf("Root certs do not match: %v", err)
-			}
+			t.LogStep("Checking response from productpage.")
+			retry.UntilSuccess(t, func(t test.TestHelper) {
+				curl.Request(t, app.BookinfoProductPageURL(t, meshNamespace), curl.NilRequestOption{}, assert.ResponseStatus(200))
+			})
 
-			log.Log.Info("Verifying the CA certificate")
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s -text -noout > %s/ca-cert.crt.txt`, sampleCACert, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			_, err = util.ShellMuteOutput(`openssl x509 -in %s/proxy-cert-2.pem -text -noout > %s/pod-cert-chain-ca.crt.txt`, tmpDir, tmpDir)
-			util.Inspect(err, "Failed to print cert", "", t)
-			if err := util.CompareFiles(fmt.Sprintf("%s/ca-cert.crt.txt", tmpDir), fmt.Sprintf("%s/pod-cert-chain-ca.crt.txt", tmpDir)); err != nil {
-				t.Errorf("CA certs do not match: %v", err)
-			}
+		} else {
 
-			log.Log.Info("Verifying the certificate chain")
-			output, err := util.ShellMuteOutput(`/bin/bash -c "openssl verify -CAfile <(cat %s %s) %s/proxy-cert-1.pem"`, sampleCACert, sampleCARoot, tmpDir)
-			util.Inspect(err, "Failed to verify the certificate chain", "", t)
-			expected := []byte(fmt.Sprintf("%s/proxy-cert-1.pem: OK", tmpDir))
-			if err := util.Compare([]byte(output), expected); err != nil {
-				t.Errorf("unexpected output while verifying cert chain: %v", err)
-			}
+			t.LogStep("Retrieving certificates.")
+			tmpDir = shell.CreateTempDir(t, "cacerts")
+			oc.Exec(t, pod.MatchingSelector("app=productpage", "bookinfo"), "istio-proxy",
+				fmt.Sprintf(`openssl s_client -showcerts -connect details:9080 > '%s/bookinfo-proxy-cert.txt' || true`, tmpDir))
+
+			t.LogStep("Extracting certificates from response.")
+			shell.Executef(t,
+				`sed -n '/-----BEGIN CERTIFICATE-----/{:start /-----END CERTIFICATE-----/!{N;b start};/.*/p}' '%s/bookinfo-proxy-cert.txt' > '%s/certs.pem'`,
+				tmpDir, tmpDir)
+			shell.Executef(t,
+				`awk 'BEGIN {counter=0;} /BEGIN CERT/{counter++} { print > "%s/proxy-cert-" counter ".pem"}' < '%s/certs.pem'`,
+				tmpDir, tmpDir)
```

Comment:
This should really be done in the code instead of calling external tools. Most people don't understand sed commands. 

Idea: ask ChatGPT to implement the equivalent go code?

Reply:
And instead of writing to a file, if we can do all this in memory, that would be great. We can probably replace the openssl calls with go functions from `crypto/x509`. Maybe ask ChatGPT something like "given two x509 certificate files, how can I compare them in go?"

Reply:
I am trying to finish a cert util tool before next week...
Yes, better to use golang https://pkg.go.dev/crypto/x509  lib.

I doubt ChatGPT may return all right answers if the user who asked the question doesn't understand what looks right or wrong ... anyway, that's off topic and it's still a good place to search.